### ISSUE TITLE
Change Request: Include `ruleId` in error message of crash

### ISSUE THREAD

**Issue by AriPerkkio** [eslint/eslint#15037]
### ESLint version

7.32.0, 8.0.0-beta.1

### What problem do you want to solve?

When ESLint rule encounters unexpected error and crashes, the error log should indicate which rule caused the crash. There is already some custom error handling done:

https://github.com/eslint/eslint/blob/143a5987f18f063a47a0646fa1e10e0f88602f6f/lib/linter/linter.js#L1213-L1226

It is important for users to identify which rule is causing linter to crash. This information is used when temporarily disabling the erroneous rule. It is also essential when reproducing and debugging the issue. Users and plugin developers are most interested in the root cause.

Most of the time the stack trace already displays the erroneous rule but there are many cases where it does not. Below are some real examples from stable releases of various eslint plugins.

<details>
    <summary>Rule code is minified</summary>

```
TypeError: Cannot read property 'type' of undefined
Occurred while linting <text>:45
    at analyzePropertyChain (/<removed>/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:2235:12)
    at analyzePropertyChain (/<removed>/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:2264:20)
    at /<removed>/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:1297:34
    at Array.forEach (<anonymous>)
    at visitFunctionWithDependencies (/<removed>/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:1277:43)
    at visitCallExpression (/<removed>/eslint-plugin-react-hooks/cjs/eslint-plugin-react-hooks.development.js:1737:11)
    at /<removed>/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/<removed>/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/<removed>/eslint/lib/linter/node-event-generator.js:254:26)
```

</details>

<details>
    <summary>Recursion fills stack trace</summary>

```
RangeError: Maximum call stack size exceeded
Occurred while linting <removed>/oaf-project/oaf-react-final-form/src/validation/index.ts:125
    at <removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:109:66
    at Array.every (<anonymous>)
    at isTypeReadonlyRecurser (<removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:109:55)
    at <removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:36:43
    at Array.some (<anonymous>)
    at checkTypeArguments (<removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:36:27)
    at isTypeReadonlyArrayOrTuple (<removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:48:16)
    at isTypeReadonlyRecurser (<removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:123:29)
    at <removed>/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:109:66
    at Array.every (<anonymous>)

```

</details>

<details>
    <summary>Rule is using utilities and third party helpers</summary>

```
TypeError: Cannot read property 'flags' of undefined
Occurred while linting /<removed>/querycap/webappkit/@querycap/reactutils/must.tsx:7
    at Object.isUnionType (/<removed>/tsutils/typeguard/2.8/type.js:69:18)
    at unionTypeParts (/<removed>/tsutils/util/type.js:92:19)
    at isPropertyReadonlyInType (/<removed>/tsutils/util/type.js:151:21)
    at isReadonlyPropertyFromMappedType (/<removed>/tsutils/util/type.js:197:12)
    at /<removed>/tsutils/util/type.js:179:21
    at someTypePart (/<removed>/tsutils/util/type.js:100:52)
    at isReadonlyPropertyIntersection (/<removed>/tsutils/util/type.js:172:12)
    at Object.isPropertyReadonlyInType (/<removed>/tsutils/util/type.js:161:43)
    at isTypeReadonlyObject (/<removed>/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:72:28)
    at isTypeReadonlyRecurser (/<removed>/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:127:30)
```

</details>

<details>
    <summary>Not sure what's going on here</summary>

```
TypeError: Cannot destructure property `object` of 'undefined' or 'null'.
Occurred while linting <text>:21
    at CallExpression[callee.property.name=/toBe(Truthy|Falsy)?|toEqual/][callee.object.callee.name='expect'] (/<removed>/ci/node_modules/eslint-plugin-jest-dom/dist/createBannedAttributeRule.js:34:21)
    at /<removed>/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/<removed>/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/<removed>/node_modules/eslint/lib/linter/node-event-generator.js:254:26)
    at NodeEventGenerator.applySelectors (/<removed>/node_modules/eslint/lib/linter/node-event-generator.js:283:22)
    at NodeEventGenerator.enterNode (/<removed>/node_modules/eslint/lib/linter/node-event-generator.js:297:14)
    at CodePathAnalyzer.enterNode (/<removed>/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:711:23)
    at /<removed>/node_modules/eslint/lib/linter/linter.js:952:32
    at Array.forEach (<anonymous>)
```

</details>

### What do you think is the correct solution?

ESLint should include the erroneous rule id in the error logs.

Example of ideal error log. Line 7 is added by this feature. Line numbers are only for demo purpose here.

```
 1 | Oops! Something went wrong! :(
 2 |
 3 | ESLint: 8.0.0-beta.1
 4 |
 5 | TypeError: Cannot read property 'flags' of undefined
 6 | Occurred while linting /<removed>/querycap/webappkit/@querycap/reactutils/must.tsx:7
 7 | Rule: "some-plugin/some-rule"
 8 |     at Object.isUnionType (/<removed>/tsutils/typeguard/2.8/type.js:69:18)
 9 |     at unionTypeParts (/<removed>/tsutils/util/type.js:92:19)
10 |     at isPropertyReadonlyInType (/<removed>/tsutils/util/type.js:151:21)
11 |     at isReadonlyPropertyFromMappedType (/<removed>/tsutils/util/type.js:197:12)
12 |     at /<removed>/tsutils/util/type.js:179:21
13 |     at someTypePart (/<removed>/tsutils/util/type.js:100:52)
14 |     at isReadonlyPropertyIntersection (/<removed>/tsutils/util/type.js:172:12)
15 |     at Object.isPropertyReadonlyInType (/<removed>/tsutils/util/type.js:161:43)
16 |     at isTypeReadonlyObject (/<removed>/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:72:28)
17 |     at isTypeReadonlyRecurser (/<removed>/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js:127:30)
```

I have required changes ready for this feature. I'll setup PR if this gets accepted.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
Thanks for the proposal, this seems like something that could be very useful :+1:

**Comment by nzakas** [comment]
I like it. :+1:

**Comment by nzakas** [comment]
Marking as accepted.
